File: src/java/spring/src/main/java/com/benchmark/service/CacheService.java
package com.benchmark.service;

import com.benchmark.model.CacheResult;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.redis.core.RedisTemplate;
import org.springframework.stereotype.Service;

import java.time.Duration;
import java.time.Instant;

@Service
public class CacheService {
    private final RedisTemplate<String, String> redisTemplate;

    @Autowired
    public CacheService(RedisTemplate<String, String> redisTemplate) {
        this.redisTemplate = redisTemplate;
    }

    public CacheResult getOrSet(String key) {
        String value = redisTemplate.opsForValue().get(key);
        if (value != null) {
            return new CacheResult(key, value, true);
        }

        // No Thread.sleep(50) on the miss path. It was an invariante 1 violation
        // (docs/ACTION_PLAN.md): hidden work on the measured path. On the
        // kotlin/spring sibling that carried the same sleep, the cache read
        // never hit, so every request paid it and /cache measured 1,962 rps
        // against the 100-connections / 50 ms ceiling of 2,000 -- the number
        // described the delay, not Redis.
        //
        // Even where the read does hit, the measurement window (30 s warmup +
        // 5 x 60 s = 330 s) outlives the 300 s TTL, so the key expires
        // mid-sequence and one repetition eats the stall.
        value = "Cached value for " + key + " at " + Instant.now().toString();
        redisTemplate.opsForValue().set(key, value, Duration.ofSeconds(300));
        return new CacheResult(key, value, false);
    }
}

File: src/kotlin/spring/src/main/kotlin/com/benchmark/service/CacheService.kt
package com.benchmark.service

import org.springframework.data.redis.core.RedisCallback
import org.springframework.data.redis.core.RedisTemplate
import org.springframework.stereotype.Service
import java.time.Duration

@Service
class CacheService(
    private val redisTemplate: RedisTemplate<String, String>
) {
    fun get(key: String): String? {
        return redisTemplate.opsForValue().get(key)
    }

    fun set(key: String, value: String, ttlSeconds: Int = 300): Boolean {
        return try {
            redisTemplate.opsForValue().set(key, value, Duration.ofSeconds(ttlSeconds.toLong()))
            true
        } catch (e: Exception) {
            println("Error setting cache key $key: ${e.message}")
            false
        }
    }

    fun getOrSet(key: String, factory: () -> String, ttlSeconds: Int = 300): Pair<String, Boolean> {
        return try {
            val existing = get(key)
            if (existing != null) {
                Pair(existing, true)
            } else {
                val value = factory()
                set(key, value, ttlSeconds)
                Pair(value, false)
            }
        } catch (e: Exception) {
            println("Error in getOrSet for key $key: ${e.message}")
            Pair(factory(), false)
        }
    }

    /// Goes through RedisTemplate.execute, which borrows a connection from the
    /// Lettuce pool and returns it.
    ///
    /// It used to call `redisTemplate.connectionFactory.connection.ping()`, which
    /// obtains a *new* connection on every invocation and never closes it. With
    /// `spring.data.redis.lettuce.pool.max-active=32` from the ConfigMap, the
    /// /health scenario drained the pool within the first few dozen requests and
    /// every later caller blocked waiting for one: /health measured 2,910 rps
    /// with an 8,018 ms p99.
    ///
    /// The damage did not stop at /health. Once the pool was exhausted, every
    /// cache read in the same pod failed too, so getOrSet always took the miss
    /// branch -- which is why /cache sat at 1,962 rps against the
    /// 100-connections / 50 ms ceiling of the Thread.sleep(50) that used to live
    /// in CacheController. One leak explained both numbers.
    fun ping(): Boolean {
        return try {
            redisTemplate.execute(RedisCallback { it.ping() }) == "PONG"
        } catch (e: Exception) {
            false
        }
    }
}
